### PR TITLE
[RFC] device property: Don't set secondary pointer if it is not NULL 

### PR DESCRIPTION
**Pull request by bardliao**
…re node pointer

If we do the following sequence of calls

	set_secondary_fwnode(dev, NULL);
	set_primary_fwnode(dev, NULL);

The secondary will be set to ERR_PTR(-ENODEV) in set_primary_fwnode(),
and it will not pass the fwnode_is_primary() test of call

set_secondary_fwnode(dev, secondary node);

The idea is that if (dev->fwnode) is not NULL, it must be primary no
matter what secondary is.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

### CODE REVIEW

File: drivers/base/core.c [hunk 1]
```diff
@@ -4300,7 +4300,7 @@ void set_secondary_fwnode(struct device *dev, struct fwnode_handle *fwnode)
 	if (fwnode)
 		fwnode->secondary = ERR_PTR(-ENODEV);
 
-	if (fwnode_is_primary(dev->fwnode))
+	if (dev->fwnode)
```
Comment (plbossart) [inline drivers/base/core.c]: The change may be right, but I don't see how this fixes the timeouts and what the relationship is between pm_runtime issues and device properties?
Also your commit message is hard to follow, can you revisit the wording and explain the sequence that leads to the problem better?

**Review by plbossart — CHANGES_REQUESTED** [review]
not fully understanding the problem and the fix?

File: drivers/base/core.c (hunk 1 above)
Comment (plbossart) [reply drivers/base/core.c]: It's also not clear to me why this change is only required *after* @andy-shev changes - Really lost on how Andy's changes introduced the resume timeouts we are seeing and how this patches corrects the behavior.

**Comment by andy-shev** [comment]
I guess here is a misunderstanding what primary and secondary means. I tried to describe this in the https://github.com/andy-shev/linux/commit/d5dcce0c414fcbfe4c2037b66ac69ea5f9b3f75c. Also there may be different cases: primary and secondary are coupled, or they are independent and they (each of them) may be exclusive or shared (in this commit https://github.com/andy-shev/linux/commit/99aed9227073fb34ce2880cbc7063e04185a65e1 i explained case when primary is shared and what is the outcome of breaking the link). Note, primary node is configured by OF or ACPI and by individual drivers by sharing it among devices. Cleaning happens only when device is gone (by device core).

Can we see better explanation what's the problem is?

**Comment by andy-shev** [comment]
Okay, we have these cases:
* primary->secondary->-ENODEV
* primary->NULL
* secondary->-ENODEV
* NULL

Last one is easy.

First one goes like this. `set_secondary_fwnode(dev, NULL)`  converts this to the second case.

Second one is relatively easy. `set_secondary_fwnode(dev, NULL)` is no op, and `set_primary_fwnode(dev, NULL)` does this `dev->fwnode = fn->secondary;` which effectivelly assigns NULL to the `dev->fwnode`.

Third one is simply check if it's secondary and assign NULL directly to `dev->fwnode`. `set_primary_fwnode()` becomes no-op.

**Comment by andy-shev** [comment]
> The secondary will be set to `ERR_PTR(-ENODEV)` in `set_primary_fwnode()`, and it will not pass the `fwnode_is_primary()` test of call  `set_secondary_fwnode(dev, secondary node);`
> 
> The idea is that if (dev->fwnode) is not NULL, it must be primary no matter what secondary is.

And the last is not correct. If you assign secondary without primary, something is not okay.

I guess what you are trying to do is actually to have

* secondary->-ENODEV

follwed by `set_secondary_fwnode(dev, secondary);`. No, you can't have two **secondary** fwnodes in a chain.

**Comment by plbossart** [comment]
@andy-shev for more context, see https://github.com/thesofproject/linux/pull/2603#issue-533132677. 

What we identified later is that the timeouts happens if we first have a sequence of modprobe/rmmod and remove/re-add devices. We bisected these timeouts to the introduction of timeouts to the two patches you added in v5.10-rc2. Reverting the 2 patches solves the timeout issue.

Here in this PR @bardliao was trying to see if there was a problem with properties, but I am with you, I don't get what the suggested fixes do nor why they would impact SoundWire suspend-resume. The one point I do get is we share the same fwnode between two layers so there's isn't a zero correlation with your changes.

**Comment by andy-shev** [comment]
@plbossart Can you do a quick test if replacing -ENODEV to NULL in that assignment (`...->secondary = ERR_PTR(-ENODEV)`) helps?

**Comment by plbossart** [comment]
> @plbossart Can you do a quick test if replacing -ENODEV to NULL in that assignment (`...->secondary = ERR_PTR(-ENODEV)`) helps?

Sorry, not following @andy-shev. The rest of the code uses IS_ERR, except where it doesn't...

 e.g.


````c
static inline bool fwnode_is_primary(struct fwnode_handle *fwnode)
{
	return fwnode && !IS_ERR(fwnode->secondary);
}

void set_secondary_fwnode(struct device *dev, struct fwnode_handle *fwnode)
{
	if (fwnode)
		fwnode->secondary = ERR_PTR(-ENODEV);

static int fwnode_property_read_int_array(const struct fwnode_handle *fwnode,
					  const char *propname,
					  unsigned int elem_size, void *val,
					  size_t nval)
{
	int ret;

	ret = fwnode_call_int_op(fwnode, property_read_int_array, propname,
				 elem_size, val, nval);
	if (ret == -EINVAL && !IS_ERR_OR_NULL(fwnode) &&
	    !IS_ERR_OR_NULL(fwnode->secondary))
...

struct fwnode_handle *device_get_next_child_node(struct device *dev,
	/* When no more children in primary, continue with secondary */
	if (fwnode && !IS_ERR_OR_NULL(fwnode->secondary))
````

**Comment by andy-shev** [comment]
Reverting only this (it needs to be done manually because of another patch) https://github.com/andy-shev/linux/commit/d5dcce0c414fcbfe4c2037b66ac69ea5f9b3f75c. I believe if it helps the problem is exactly what I described: an attempt to assign a secondary node to device which already has one.

**Comment by bardliao** [comment]
Thanks @andy-shev for the explanation.

> Reverting only this (it needs to be done manually because of another patch) [andy-shev@d5dcce0](https://github.com/andy-shev/linux/commit/d5dcce0c414fcbfe4c2037b66ac69ea5f9b3f75c). I believe if it helps the problem is exactly what I described: an attempt to assign a secondary node to device which already has one.

Yes, it helps. 


> I guess what you are trying to do is actually to have
> 
> * secondary->-ENODEV
> 
> follwed by `set_secondary_fwnode(dev, secondary);`. No, you can't have two **secondary** fwnodes in a chain.

You are right. But the `secondary` here is configured by ACPI and it was `primary->secondary->-ENODEV`. It became `secondary->-ENODEV`after `set_secondary_fwnode(dev, NULL);` and `set_primary_fwnode(dev, NULL)`; . Below is the trace log that I currently have. I will get more trace and explain it in more detail latter.
```
[    5.282230] rt711 sdw:1:25d:711:0: set_secondary_fwnode start dev->fwnode 00000000e7090087 dev->fwnode->secondary 0000000000000000 fwnode 0000000088f1f67d
[    5.282232] rt711 sdw:1:25d:711:0: set_secondary_fwnode done dev->fwnode 00000000e7090087 dev->fwnode->secondary 0000000088f1f67d
[   29.475313] rt711 sdw:1:25d:711:0: set_secondary_fwnode start dev->fwnode 00000000e7090087 dev->fwnode->secondary 0000000088f1f67d fwnode 0000000000000000
[   29.475316] rt711 sdw:1:25d:711:0: set_secondary_fwnode done dev->fwnode 00000000e7090087 dev->fwnode->secondary 0000000000000000
[   29.525724] soundwire sdw:1:25d:711:0: set_primary_fwnode start dev->fwnode 00000000e7090087 dev->fwnode->secondary 0000000000000000 fwnode 0000000000000000
[   29.525726] soundwire sdw:1:25d:711:0: set_primary_fwnode done dev->fwnode 0000000000000000
[   31.654916] rt711 sdw:1:25d:711:0: set_secondary_fwnode start dev->fwnode 00000000e7090087 dev->fwnode->secondary ffffffffffffffed fwnode 000000008cd3755f
[   31.654918] rt711 sdw:1:25d:711:0: set_secondary_fwnode done dev->fwnode 000000008cd3755f dev->fwnode->secondary ffffffffffffffed
[   39.836593] rt711 sdw:1:25d:711:0: set_secondary_fwnode start dev->fwnode 00000000e7090087 dev->fwnode->secondary ffffffffffffffed fwnode 000000004a1bf865
[   39.836595] rt711 sdw:1:25d:711:0: set_secondary_fwnode done dev->fwnode 000000004a1bf865 dev->fwnode->secondary ffffffffffffffed
```

**Comment by andy-shev** [comment]
> > I guess what you are trying to do is actually to have
> > 
> > * secondary->-ENODEV
> > 
> > follwed by `set_secondary_fwnode(dev, secondary);`. No, you can't have two **secondary** fwnodes in a chain.
> 
> You are right. But the `secondary` here is configured by ACPI and it was `primary->secondary->-ENODEV`.

ACPI doesn't set secondary. Something is way wrong in your case if it's true. Custom patches?

> It became `secondary->-ENODEV`after `set_secondary_fwnode(dev, NULL);` and `set_primary_fwnode(dev, NULL)`; . Below is the trace log that I currently have. I will get more trace and explain it in more detail latter.

I'm sorry I barely understand what is going on in this log. The rule of thumb is when you provide a trace, provide a code which gives that trace to understand where it's exactly taken and what is the mean of the fields in output.

**Comment by plbossart** [comment]
@andy-shev @bardliao I added a trace of what I think goes South, see https://github.com/plbossart/sound/commit/76d0f5a8ffe481cce442e6e47fb5e6570a081372

I think when we unload a driver and call set_primary_fwnode(NULL), the fwnode remains secondary (set to ERR_PTR(-NODEV)),  so the next time we probe and add properties we end-up overriding the fwnode with the properties. 

There's something missing in the initialization, either in the device core or the SoundWire core, so that on probe the previous fwnode->secondary has to become NULL again.

**Comment by bardliao** [comment]
@andy-shev @plbossart If we use `set_primary_fwnode` to remove a `primary->NULL` fwnode, the `primary` pointer will end up becoming primary-> ERR_PTR(-ENODEV). And it will be treated as a secondary fwnode.
See the code in `set_primary_fwnode` below
```
struct fwnode_handle *fn = dev->fwnode;   <<<< say fn is (A)
...
        } else {
                if (fwnode_is_primary(fn)) {
                        dev->fwnode = fn->secondary; <<<< dev->fwnode = NULL which is expected
                        if (!(parent && fn == parent->fwnode) && fn->secondary)
                                fn->secondary = ERR_PTR(-ENODEV); <<<< fn (A) -> secondary is now ERR_PTR(-ENODEV); which seems to be incorrect. 
                } else {
...
```
The updated commit suggested a fix. Does it make sense?

**Review by andy-shev — COMMENTED** [review]
No. This makes no sense. NAK.

**Comment by andy-shev** [comment]
> @andy-shev @plbossart If we use `set_primary_fwnode` to remove a `primary->NULL` fwnode, the `primary` pointer will end up becoming primary-> ERR_PTR(-ENODEV). And it will be treated as a secondary fwnode.

The above makes no sense. I think you didn't get how it supposed to work. When you call `set_primary_fwnode(dev, NULL);` against `primary --> NULL` chain, it effectively means `dev->fwnode = NULL;`.

> The updated commit suggested a fix. Does it make sense?

No.

As I told there is a mix of terminology and position in the list. `struct fwnode_handle` defines a pointer to the next element in a single-linked list. The list is restricted to have at least 0 at most 2 elements. Possibilities are:
* NULL
* primary --> NULL
* secondary --> -ENODEV
* primary --> secondary --> -ENODEV

All other combinations are illegal!

**Comment by andy-shev** [comment]
> @andy-shev @bardliao I added a trace of what I think goes South, see [plbossart@76d0f5a](https://github.com/plbossart/sound/commit/76d0f5a8ffe481cce442e6e47fb5e6570a081372)
> 
> I think when we unload a driver and call set_primary_fwnode(NULL), the fwnode remains secondary (set to ERR_PTR(-NODEV)), so the next time we probe and add properties we end-up overriding the fwnode with the properties.
> 
> There's something missing in the initialization, either in the device core or the SoundWire core, so that on probe the previous fwnode->secondary has to become NULL again.

I will put here your words from above link:
> And on the second insert, the dev->fwnode will be modified - really nasty side effect. On a device remove, we really don't want the fwnode secondary type to be persistent....

Hooray!  Now you understand the issue with `device_add_properties()` and why we insist to use swnode directly. Your code is responsible to allocate / register / unregister swnodes at exact moments when it's appropriate from **your code** perspective.

**UPDATE** `device_*_properties()` compounds together alloc+register+bind vs. unregister+release+unbind. And it **does not** work for the case
* secondary --> -ENODEV

**Comment by bardliao** [comment]
> > @andy-shev @plbossart If we use `set_primary_fwnode` to remove a `primary->NULL` fwnode, the `primary` pointer will end up becoming primary-> ERR_PTR(-ENODEV). And it will be treated as a secondary fwnode.
> 
> The above makes no sense. I think you didn't get how it supposed to work. When you call `set_primary_fwnode(dev, NULL);` against `primary --> NULL` chain, it effectively means `dev->fwnode = NULL;`.

@andy-shev Yes, `dev->fwnode = NULL;`, but my point is about the original dev->fwnode pointer which we removed, but will be reused when the driver is loaded again.
```
[   22.165910] soundwire sdw:3:25d:714:1: bard: set_primary_fwnode start dev->fwnode 0000000094d3ce7a fwnode 0000000000000000 <<<< We call set_primary_fwnode(dev, NULL) to remove a primary->NULL fwnode
[   22.165913] soundwire sdw:3:25d:714:1: bard: set_primary_fwnode !(parent && fn == parent->fwnode)		<<<< fn->secondary is set to ERR_PTR(-ENODEV) with the if (!(parent && fn == parent->fwnode)) condition
[   22.165915] soundwire sdw:3:25d:714:1: bard: set_primary_fwnode parent sdw-master-3 parent->fwnode 00000000b2414cc2 dev->fwnode 0000000000000000 
[   22.165918] soundwire sdw:3:25d:714:1: bard: set_primary_fwnode done dev->fwnode 0000000000000000		<<<< dev->fwnode is NULL which is expected, but at the same time the original fwnode pointer (0000000094d3ce7a) ->secondary is ERR_PTR(-ENODEV)
...
[   24.319843] soundwire sdw:3:25d:714:1: bard: set_primary_fwnode start dev->fwnode 0000000000000000 fwnode 0000000094d3ce7a
[   24.319844] soundwire sdw:3:25d:714:1: bard: set_primary_fwnode done dev->fwnode 0000000094d3ce7a
[   24.319845] soundwire sdw:3:25d:714:1: bard: set_primary_fwnode dev->fwnode->secondary ffffffffffffffed	<<<< We call set_primary_fwnode(dev, fwnode), but it end up with primary->ERR_PTR(-ENODEV)
```
[dmesg.txt](https://github.com/thesofproject/linux/files/5695610/dmesg.txt)
The trace commit is https://github.com/bardliao/linux/commit/ce4a7e48e41170e07669f81fdc1d23952192450c
And the testing tree is https://github.com/bardliao/linux/tree/fwnode-trace

> 
> > The updated commit suggested a fix. Does it make sense?
> 
> No.
> 
> As I told there is a mix of terminology and position in the list. `struct fwnode_handle` defines a pointer to the next element in a single-linked list. The list is restricted to have at least 0 at most 2 elements. Possibilities are:
> 
> * NULL
> * primary --> NULL
> * secondary --> -ENODEV
> * primary --> secondary --> -ENODEV
> 
> All other combinations are illegal!

Yes, this PR is exactly trying to avoid the unexpected * primary --> -ENODEV
The issue what I see is that fwnode(A) will become a secondary fwnode if we call set_primary_fwnode in below sequence

* step 1: set_primary_fwnode(dev, A)
    * result 1: dev->fwnode = A->NULL
* step 2: set_primary_fwnode(dev, NULL)
    * result 2: dev->fwnode = NULL
* step 3: set_primary_fwnode(dev, A)
    * result 3: dev->fwnode = A->-ENODEV
Or maybe we should set A->NULL before step 3?

**Comment by plbossart** [comment]
> Hooray! Now you understand the issue with `device_add_properties()` and why we insist to use swnode directly. Your code is responsible to allocate / register / unregister swnodes at exact moments when it's appropriate from **your code** perspective.
> 
> **UPDATE** `device_*_properties()` compounds together alloc+register+bind vs. unregister+release+unbind. And it **does not** work for the case
> 
> * secondary --> -ENODEV

@andy-shev we do remove the properties when appropriate without relying on the device core. we did this a while ago. Even if we used swnodes directly, the problem would still exist:

The problem is that when doing an rmmod snd-soc-rt711, ACPI_COMPANION_SET is called with a NULL parameter, so the fwnode is tagged as secondary. Then when we redo a modprobe snd-sof-rt711 and later add properties, the fwnode is not considered as primary, so the dev->fwnode will be overridden by the properties. In other words, we get a different result during the second iteration of our kmod load/unload tests.

I think it's keeping the type of the fwnode that is problematic when dealing with probe/remove. That's what @bardliao suggested forcing the fwnode->secondary= NULL to force the fwnode back to a primary when registering a device.

**Comment by plbossart** [comment]
let's follow the review with #2625 or #2635

**Comment by andy-shev** [comment]
> The problem is that when doing an rmmod snd-soc-rt711, ACPI_COMPANION_SET is called with a NULL parameter, so the fwnode is tagged as secondary.

So, from the state of `primary --> secondary --> -ENODEV` it comes to `secondary --> -ENODEV`. This is correct.
The word `tagget` above confuses me. What does it mean?

> Then when we redo a modprobe snd-sof-rt711 and later add properties, the fwnode is not considered as primary, so the dev->fwnode will be overridden by the properties. In other words, we get a different result during the second iteration of our kmod load/unload tests.

Right.

> I think it's keeping the type of the fwnode that is problematic when dealing with probe/remove. That's what @bardliao suggested forcing the fwnode->secondary= NULL to force the fwnode back to a primary when registering a device.

Okay, but do you agree that this is not the best approach?
